### PR TITLE
Split classic boot configuration

### PR DESCRIPTION
**Pull request by waveform80**
This is a fairly complex PR that moves the classic branch closer to the core branch in preparation for eventual merging (although this is not yet that merge as other things need to happen in snapd and flash-kernel for that to occur), and splits the boot configuration for easier maintenance. Specifically, this PR:

* Splits the boot configuration into a series of configuration files, most of which are not (intended to be) user-editable
* Instead of copying a single u-boot binary to the boot partition, copies all u-boot binaries from the u-boot package under different filenames (which the new config.txt will select according to platform); this is how core already does things (albeit from a different source, as core builds its own u-boot binaries)
* Uses the boot.scr uboot script from the flash-kernel package instead of a (redundant) one that was included in this branch (this is another move in preparation for merging with core as the pi boot script in flash-kernel will eventually support both platforms)
* Includes the option to use a different series than the builder's for packages

For review, I would recommend going through the commits individually then looking at the end state rather than trying to review the overall diff.

### CODE REVIEW

File: Makefile
Comment (sil2100) [inline Makefile:58]: Good to see we'll be finally pulling the bootscript from a common place (flash-kernel). Just a note here: before we can actually merge this, we'll probably need to SRU flash-kernel to bionic, since currently it doesn't have the bootscr.rpi in it. Even though the next bionic point release is still in some time, we do build dailies for that series. So I'd say we'd want flash-kernel in at least bionic-proposed.

File: configs/syscfg.txt
Comment (sil2100) [inline configs/syscfg.txt:9]: Ah, so this is the place which the bluetooth package will modify to switch bluetooth on and off, right?

File: configs/README
Comment (sil2100) [inline configs/README:26]: Love this documentation - this is something we've been missing even before. Thanks for adding it!

**Review by sil2100 — APPROVED** [review]
+1 on the changes. I will not merge it yet as per the inline comment regarding flash-kernel. But other than that, this looks like good to go (as I'm sure you tested this on a live system).

File: configs/syscfg.txt
Comment (waveform80) [reply configs/syscfg.txt:9]: This is indeed the intention; flip this to `btcfg.txt` and back to switch between the configs (and eventually we'll have a CLI tool to handle this and the manipulation of the dtparam lines above).

File: Makefile
Comment (waveform80) [reply Makefile:58]: Actually I don't think this is an issue. There's a very subtle change in the script which deals with this different. Specifically, Eoan (and Disco) have bootscript/all/bootscr.rpi in flash-kernel, while Bionic has bootscript/all/bootscr.rpi3 (and a uboot.env for pi2) in its flash-kernel. Now, on line 66 note that mkimage is called with `.../bootscr.rpi*` rather than `.../bootscr.rpi` to account for this difference (yes, it's a bit hacky to use a wildcard here but I think it's acceptable given the requirement).

Incidentally, this is (one of the reasons) why I suggested going through the individual commits rather than just looking at the end state; 17596c4f3458b4e368224453f1defe0ce2bfaa69 mentions this case in the commit message.

Comment (sil2100) [reply Makefile:58]: hm, you are right and flash-kernel from bionic does have the base bootscript for the pi3. But I still think there will be a problem right now - a bit of a different one though. So the builds will not FTBFS, but I suppose our arm64 bionic daily images will stop working.

The old flash-kernel bootscr.rpi3 that we have there I guess is very armhf oriented (which is why it's in 'armhf' instead of 'all'), supporting only booting the armhf kernel - since there are no conditional checks there for now. Certainly this is less of a blocker, as we never really give any guarantee of bootability of our daily images, but it's something we should sort out earlier before the point release. So I'd say let's aim for getting it into bionic-proposed at least in the nearest time.

**Comment by mvo5** [comment]
A bit of a meta-question: once we are happy with things here we will port to "pi2-gadget", "cm3-gadget" and "pi-gadget" (the  last one is the unified pi gadget we use in core18) ?

**Comment by waveform80** [comment]
> A bit of a meta-question: once we are happy with things here we will port to "pi2-gadget", "cm3-gadget" and "pi-gadget" (the last one is the unified pi gadget we use in core18) ?

Well, this PR is only a step towards unification of the classic and core gadgets; there's a bit more work to be done (including within snapd itself) before that can be achieved. If pi-gadget is the unified one it may be worth continuing the effort there (on the core side of things); whether it's worth merging further work into pi2-gadget and cm3-gadget I don't know. My understanding was that the eventual plan was to retire those in favour of a unified image?

**Pull request merged**
